### ISSUE TITLE
twisted.conch.ssh.channel has poor tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2709]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| z3p reported|
|-|-|
|Trac ID|trac#2709|
|Type|defect|
|Created|2007-06-19 20:35:46Z|
|Branch|https://github.com/twisted/twisted/tree/channel-2709|

See [#2678](https://github.com/twisted/twisted/issues/2678), [#2679](https://github.com/twisted/twisted/issues/2679), [#2682](https://github.com/twisted/twisted/issues/2682), [#2707](https://github.com/twisted/twisted/issues/2707).

<details><summary>Searchable metadata</summary>

```
trac-id__2709 2709
type__defect defect
reporter__z3p z3p
priority__highest highest
milestone__ 
branch__branches_channel_2709 branches/channel-2709
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__soc2007 soc2007
time__1182285346000000 1182285346000000
changetime__1203453164000000 1203453164000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p removed owner|
|-|-|

Ready for review in channel-2709.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to z3p|
|-|-|

Most of the docstrings for the new tests don't really say what they're testing.  "works correctly", "returns an appopriate value", "handles data appropriately", and so on, aren't telling me what the expected results are.  For example, in the case of "str", you should say something about what is expected to show up in the str's results.  The docstring can then be something like "str(SSHChannel) should return a string describing the channel's foo, bar, and baz attributes at a glance."

Missing docstrings entirely:
  * ```test_loseConnection```
  * ```test_getPeer```
  * ```test_getHost```
  * ```setUp``` (why is ```remoteMaxPacket```1. is that important to the tests?  why?)

Docstrings are always required, but they are particularly important in this code because it has historically proven particularly difficult for other maintainers to understand.  Test and module docstrings are a good place to provide an overview of the intended functionality of the system rather than the snapshots 

The tests' implementations doesn't help me understand the vague docstrings, either.  For example, ```test_init```:  131072?  32768?  10?  1, 2, 3, 4, 5, 6, 7?  What do any of these things mean?  Use some named constants if the values are meaningful, and if not, then comment to that effect.

Since it's really verifying the behavior of some stuff that's going on in ```setUp```, you might want to put ```test_init``` specifically on a separate class so that it can set up its own channel inline with the test.

In ...```ssh.channel``` itself:
  * don't ever comment out code.  Just delete it.  Version control can tell us about it if it's interesting.
  * I'm a bit confused as to the purpose of the changes, but hopefully that will be clearer when I can grok what's going on in the tests.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|z3p removed owner|
|-|-|

I've tried to add more information to the docstring to answer your questions, and explain the code better in t.c.s.channel.  I can't judge how I've done because the code makes sense to me.  I'm putting it back in the review category.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to z3p|
|-|-|

channel.py:
 * Unused import context
 * update its copyright
 * ```SSHChannel``` could declare its implementation of ```ITransport```.
 * there are one ```-=``` and one ```+=``` which need some spaces around

test_channel.py:
 * don't use callbacks named '_', a small meaningful name is better
 * in test_addWindowBytes, you probably want to do ```self.assertTrue(cb[0])``` not ```self.assertTrue(_)```
 * test_init should definitely be in another TestCase


Both files would benefit from a few blank lines.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|z3p commented|
|-|-|

I've fixed all but test_init.  I'm not sure it deserves its own TestCase; after all, it is testing a function of SSHChannel.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|z3p removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to z3p|
|-|-|

  * in the tests:
    * Make new classes new-style
    * Document the ivars of MockConnection and its sendClose method
  * in the implementation:
    * SSHChannel could use a class docstring with @ivars

The test method docstrings look good to me.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|z3p removed owner|
|-|-|

Okay, back up for review (channel-2709).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to z3p|
|-|-|

MockTransport's docstring should be more specific about what role the class is intended to fill.  Is it a mock for twisted.conch.ssh.transport.SSH{Client,Server}Transport?  If so, say that.  "Transport" could mean a lot of things.

Everything else looks good.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p commented|
|-|-|

(In [22472]) More documentation for MockTransport

Refs [#2709](https://github.com/twisted/twisted/issues/2709)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|z3p removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to z3p|
|-|-|

Great. Please update copyrights of both files and merge.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|z3p set status to `closed`|
|-|-|

(In [22626]) merge channel-2709

Author: z3p
Reviewer: exarkun, glyph, therve
Fixes [#2709](https://github.com/twisted/twisted/issues/2709)

Adds explicit tests and some documentation for the code in
twisted.conch.ssh.channel.  This work was done as part of Google's
Summer of Code 2007.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
